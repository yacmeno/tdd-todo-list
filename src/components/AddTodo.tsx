import React from "react";
import { ITodo } from "../hooks/useTodos";

interface IAddTodoProps {
	onAdd: (todo: ITodo) => void;
}
interface IAddTodoState {
	currentId: number;
	currentTodo: string;
	isInvalid: boolean;
}

export class AddTodo extends React.Component<IAddTodoProps, IAddTodoState> {
	readonly state: IAddTodoState = {
		currentId: 0,
		currentTodo: "",
		isInvalid: true,
	};

	private onInput(e: React.ChangeEvent<HTMLInputElement>) {
		const invalidInput = e.target.value.trim() === "";
		this.setState({
			currentTodo: e.target.value,
			isInvalid: invalidInput ? true : false,
		});
	}

	private handleReturnKey(e: React.KeyboardEvent<HTMLInputElement>) {
		if (this.state.isInvalid) {
			return;
		}

		if (e.key === "Enter") {
			this.handleSubmit();
		}
	}

	private handleSubmit() {
		this.props.onAdd({
			id: this.state.currentId,
			text: this.state.currentTodo,
			isDone: false,
		});
		this.setState(prevState => {
			return {
				currentId: prevState.currentId + 1,
				currentTodo: "",
				isInvalid: true,
			};
		});
	}

	render() {
		const { currentTodo, isInvalid } = this.state;

		return (
			<div className="todo__header">
				<input
					type="text"
					onChange={this.onInput.bind(this)}
					onKeyDown={this.handleReturnKey.bind(this)}
					value={currentTodo}
				/>
				<button
					type="button"
					disabled={isInvalid}
					onClick={() => this.handleSubmit()}
				>
					Add Todo
				</button>
			</div>
		);
	}
}
