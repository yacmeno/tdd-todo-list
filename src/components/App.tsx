import React from "react";
import useTodos, { ITodo } from "../hooks/useTodos";

const TodoApp: React.FC = () => {
	const [todos, setTodos] = useTodos();

	const dummyTodos = [
		{ id: 1, text: "Cook chicken", isDone: true },
		{ id: 2, text: "Travel to space", isDone: false },
		{ id: 3, text: "Come back to earth", isDone: false },
	];

	function clickHandler() {
		return;
	}

	return (
		<div className="app__container">
			<AddTodo />
			<Nav />
			<TodoList todos={dummyTodos} clickHandler={clickHandler} />
		</div>
	);
};

const AddTodo: React.FC = () => {
	return <div>Header</div>;
};
const Nav: React.FC = () => {
	return <div>Nav</div>;
};

export interface ITodoListProps {
	todos: ITodo[];
	clickHandler: () => void;
}

export const TodoList: React.FC<ITodoListProps> = ({ todos, clickHandler }) => {
	if (todos.length === 0) {
		return <p data-test-no-items="true">No items here!</p>;
	}

	return (
		<ul className="todo__items">
			{todos.map(todo => (
				<TodoItem key={todo.id} todo={todo} clickHandler={clickHandler} />
			))}
		</ul>
	);
};

interface ITodoItemProps {
	todo: ITodo;
	clickHandler: () => void;
}

export const TodoItem: React.FC<ITodoItemProps> = ({ todo, clickHandler }) => {
	const className = `todo__item todo__item${todo.isDone ? "--done" : ""}`;

	return (
		<li className={className} onClick={clickHandler}>
			{todo.text}
		</li>
	);
};

export default TodoApp;
