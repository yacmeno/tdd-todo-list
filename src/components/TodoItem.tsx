import React from "react";
import { ITodo } from "../hooks/useTodos";

interface ITodoItemProps {
	todo: ITodo;
	clickHandler: (todo: ITodo) => void;
}

export const TodoItem: React.FC<ITodoItemProps> = ({ todo, clickHandler }) => {
	const className = `todo__item ${todo.isDone ? "todo__item--done" : ""}`;

	return (
		<li className={className} onClick={() => clickHandler(todo)}>
			{todo.text}
		</li>
	);
};
