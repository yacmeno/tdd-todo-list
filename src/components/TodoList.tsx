import React from "react";
import { ITodo } from "../hooks/useTodos";
import { TodoItem } from "./TodoItem";

export interface ITodoListProps {
	todos: ITodo[];
	clickHandler: (todo: ITodo) => void;
}

export const TodoList: React.FC<ITodoListProps> = ({ todos, clickHandler }) => {
	if (todos.length === 0) {
		return <p data-test-no-items="true">No items here!</p>;
	}

	return (
		<ul className="todo__items">
			{todos.map(todo => (
				<TodoItem key={todo.id} todo={todo} clickHandler={clickHandler} />
			))}
		</ul>
	);
};
