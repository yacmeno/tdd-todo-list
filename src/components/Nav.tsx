import React from "react";
import { Filters } from "./TodoApp";

interface INavProps {
	activeFilter: string;
	setFilter: (filter: string) => void;
}

export const Nav: React.FC<INavProps> = ({ activeFilter, setFilter }) => {
	return (
		<ul className="navbar">
			<li
				className={activeFilter === Filters.SHOW_ALL ? "active" : ""}
				onClick={() => setFilter(Filters.SHOW_ALL)}
			>
				All
			</li>
			<li
				className={activeFilter === Filters.SHOW_TODO ? "active" : ""}
				onClick={() => setFilter(Filters.SHOW_TODO)}
			>
				Todo
			</li>
			<li
				className={activeFilter === Filters.SHOW_DONE ? "active" : ""}
				onClick={() => setFilter(Filters.SHOW_DONE)}
			>
				Done
			</li>
		</ul>
	);
};
