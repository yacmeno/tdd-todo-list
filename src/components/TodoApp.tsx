import React from "react";
import useTodos, { ITodo, ADD_TODO, TOGGLE_TODO } from "../hooks/useTodos";
import { AddTodo } from "./AddTodo";
import { TodoList } from "./TodoList";
import { Nav } from "./Nav";

const TodoApp: React.FC = () => {
	const [todos, setTodos] = useTodos();

	function toggleTodo(todo: ITodo) {
		setTodos({ type: TOGGLE_TODO, payload: todo });
	}

	function onAdd(todo: ITodo) {
		setTodos({ type: ADD_TODO, payload: todo });
	}

	return (
		<div className="app__container">
			<AddTodo onAdd={onAdd} />
			<Nav />
			<TodoList todos={todos} clickHandler={toggleTodo} />
		</div>
	);
};

export default TodoApp;
