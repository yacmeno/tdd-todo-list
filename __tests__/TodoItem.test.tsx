import React from "react";
import { shallow } from "enzyme";
import renderer from "react-test-renderer";
import { TodoItem } from "../src/components/TodoItem";

describe("<TodoItem />", () => {
	test("Should display a todo item", () => {
		const todo = { id: 0, text: "Cook something good", isDone: false };
		const clickHandler = jest.fn();

		const component = shallow(
			<TodoItem todo={todo} clickHandler={clickHandler} />
		);

		expect(component.text()).toBe(todo.text);
	});

	test("Should simulate click event", () => {
		const todo = { id: 0, text: "Cook something good", isDone: false };
		const clickHandler = jest.fn();

		const component = shallow(
			<TodoItem todo={todo} clickHandler={clickHandler} />
		);
		component.simulate("click");

		expect(clickHandler).toBeCalledTimes(1);
	});

	test("Should have different className if item isDone", () => {
		const todo = { id: 0, text: "Cook something good", isDone: true };
		const clickHandler = jest.fn();

		const componentTree = renderer
			.create(<TodoItem todo={todo} clickHandler={clickHandler} />)
			.toJSON();

		expect(componentTree).toMatchSnapshot();
	});
});
