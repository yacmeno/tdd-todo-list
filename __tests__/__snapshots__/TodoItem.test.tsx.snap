// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TodoItem /> Should have different className if item isDone 1`] = `
<li
  className="todo__item todo__item--done"
  onClick={[Function]}
>
  Cook something good
</li>
`;
